### ISSUE TITLE
Matmul won't check operand device and cause wired bugs

### ISSUE THREAD

**Issue by imyhxy** [pytorch/pytorch#49531]
## 🐛 Bug

The matrix multiplication operation of pytorch does not check whether the device of the operands are the same. If two matrices from different devices are multiplied, some subsequent operations will report errors.

## To Reproduce

1. Create two random matries, one device is CPU, another is CUDA.
1. Perform matmul operation on this two matries.
1. Print the matrix in CUDA.

## Expected behavior

Can't create CUDA tensor from python literal.
```python
import torch

# var_a in cuda, var_b in CPU
a = torch.randn((2, 2), device=torch.device('cuda'))
b = torch.randn((2, 3), device=torch.device('cpu'))

# matrix multiplication is fine
c = a @ b
d = torch.matmul(a, b)

# use torch built-in function to create tensor on CUDA is also fine
good_var = torch.randn((1, 10), device=torch.device('cuda'))

# but the following operations will failed
bad_var = torch.tensor((1, 10), device=torch.device('cuda'))
```

Trackback message:

```bash
Traceback (most recent call last):
  File "/home/fkwong/workspace/tracking/DeepSORTYolov5/deepsort/sort/torch_matmul_bug.py", line 18, in <module>
    bad_var = torch.tensor((1, 10), device=torch.device('cuda'))
RuntimeError: CUDA error: an illegal memory access was encountered
```
Also, can't print previous variable

```python
import torch

# var_a in cuda, var_b in cpu
a = torch.randn((2, 2), device=torch.device('cuda'))
b = torch.randn((2, 3), device=torch.device('cpu'))

# matrix multiplication is fine
c = a @ b
d = torch.matmul(a, b)

# use torch built-in function to create tensor on cuda is also fine
good_var = torch.randn((1, 10), device=torch.device('cuda'))

# but the following operations will failed
# bad_var = torch.tensor((1, 10), device=torch.device('cuda'))

# also won't be able to asses var_a against
print(a)
```

Trackback message:

```bash
Traceback (most recent call last):
  File "/home/fkwong/workspace/tracking/DeepSORTYolov5/deepsort/sort/torch_matmul_bug.py", line 21, in <module>
    print(a)
  File "/home/fkwong/miniconda3/envs/deepsort/lib/python3.7/site-packages/torch/tensor.py", line 153, in __repr__
    return torch._tensor_str._str(self)
  File "/home/fkwong/miniconda3/envs/deepsort/lib/python3.7/site-packages/torch/_tensor_str.py", line 371, in _str
    return _str_intern(self)
  File "/home/fkwong/miniconda3/envs/deepsort/lib/python3.7/site-packages/torch/_tensor_str.py", line 351, in _str_intern
    tensor_str = _tensor_str(self, indent)
  File "/home/fkwong/miniconda3/envs/deepsort/lib/python3.7/site-packages/torch/_tensor_str.py", line 241, in _tensor_str
    formatter = _Formatter(get_summarized_data(self) if summarize else self)
  File "/home/fkwong/miniconda3/envs/deepsort/lib/python3.7/site-packages/torch/_tensor_str.py", line 89, in __init__
    nonzero_finite_vals = torch.masked_select(tensor_view, torch.isfinite(tensor_view) & tensor_view.ne(0))
RuntimeError: copy_if failed to synchronize: cudaErrorIllegalAddress: an illegal memory access was encountered
```
## Environment

```bash
PyTorch version: 1.6.0
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: 10.0.0-4ubuntu1 
CMake version: Could not collect

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce RTX 2060
Nvidia driver version: 455.38
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.4
[pip3] torch==1.6.0
[pip3] torchreid==1.3.3
[pip3] torchvision==0.7.0
[conda] blas                      1.0                         mkl    defaults
[conda] cudatoolkit               10.2.89              hfd86e86_1    defaults
[conda] mkl                       2020.2                      256    defaults
[conda] mkl-service               2.3.0            py37he904b0f_0    defaults
[conda] mkl_fft                   1.2.0            py37h23d657b_0    defaults
[conda] mkl_random                1.1.1            py37h0573a6f_0    defaults
[conda] numpy                     1.19.4                   pypi_0    pypi
[conda] numpy-base                1.19.2           py37hfa32c7d_0    defaults
[conda] pytorch                   1.6.0           py3.7_cuda10.2.89_cudnn7.6.5_0    pytorch
[conda] torch                     1.4.0                    pypi_0    pypi
[conda] torchreid                 1.3.3                     dev_0    <develop>
[conda] torchvision               0.5.0                    pypi_0    pypi
```

cc @ngimel

**Comment by ptrblck** [comment]
Could you update to the latest stable release (`1.7.1`) and rerun your code, as I'm getting a proper error:
```python
Traceback (most recent call last):
  File "tmp.py", line 8, in <module>
    c = a @ b
RuntimeError: Tensor for argument #3 'mat2' is on CPU, but expected it to be on GPU (while checking arguments for addmm)
```

**Comment by imyhxy** [comment]
@ptrblck 
Great, I'll give it a try.

**Comment by ngimel** [comment]
This is fixed on master in 1.7, please reopen if it reproduces.
